### ISSUE TITLE
EnhancedBrowserWindowOpener is not working in Shortcut Listener

### ISSUE THREAD

**Issue by karthikeyanStar** [mcollovati/enhanced-window-opener#9]
Hi Marco,

This is karthikeyan, currently i am facing issue in EnhancedBrowserWindowOpener  when i used with shorcutlistener. I am using UI class in EnhancedBrowserWindowOpener . Why EnhancedBrowserWindowOpener is not displaying UI on shourcut listener? Can you explain ??? I am using vaadin 7.5.8... I have used EnhancedBrowserWindowOpener version 0.1 addon,
Waiting for your reply... This is very urgent......

Thanks,

Karthikeyan

**Comment by mcollovati** [comment]
Hi Karthikeyan,
can you post some code I can use to replicate the problem?
This will help a lot to fix the issue

**Comment by mcollovati** [comment]
The following code works for me

```java
Button button = new Button("Open another UI");
button.setClickShortcut(ShortcutAction.KeyCode.E, ShortcutAction.ModifierKey.CTRL, ShortcutAction.ModifierKey.SHIFT);

EnhancedBrowserWindowOpener opener = new EnhancedBrowserWindowOpener(PopupUI.class)
            .popupBlockerWorkaround(true)
            .doExtend(button);

button.addClickListener(event -> opener.open());
```
Important things here are the  **popupBlockerWorkaround=true** flag and the call to **open** inside the button click listener.
Let me know if this works also for your use case

**Comment by karthikeyanStar** [comment]
Hi Marco,

Thanks for the reply marco. I have tried the steps, but it is not working. I will explain my requirement to
you.

1. First Create a view page or window.
2. Create a button, i.e [Button button = new Button("Open another UI");]

3. Create the EnhancedBrowserWindowOpener 
i.e  [EnhancedBrowserWindowOpener opener = new EnhancedBrowserWindowOpener(PopupUI.class));]

4. Attach/Extend  your EnhancedBrowserWindowOpener  to the button.
 i.e  [opener .popupBlockerWorkaround(true).doExtend(button);]

5. Now add clickListener to button
i.e [button.addClickListener(event -> opener.open());]

6. Now you add a shortcutlistener to Window/View page, not to the button.
int[] modifierArr = new int[]{ModifierKey.CTRL,ModifierKey.ALT};
 ShortcutListener listener = new ShortcutListener("Test", KeyCode.NUM1, modifierArr) {
				 private static final long serialVersionUID = 1L;
				 @Override
				 public void handleAction(Object sender, Object target) {
					
				 }
			 };

7. On calling the shortcut key handleAction() will be invoked and so Inside the handleAction you
have to call the button.click(). This will invoke the clicklistener() of the button and must display the UI.

8. For me on invoking the shortcut keys handleAction is called and button clicklistener is also getting
invoked, but the UI is not getting displayed.

If you need additional info, please revert back.

Thanks,

Karthikeyan.

**Comment by mcollovati** [comment]
I followed the steps you describe and triggering the `EnhancedWindowOpener.open` or `Button.click`  method inside `handleAction()` seems to work but, if not running on localhost the popup is blocked by the browser (at least from chrome).
The problem here is that the when that the shortcut action is execute completely server side whereas the `popupBlockerWorkaround` works on client side.

I will try to check if there's a way to intercept the *open window* action on client side in order to simulate the *popupBlockerWorkaround* behaviour

**Comment by karthikeyanStar** [comment]
Hi Marco,

Any update on this ????

**Comment by mcollovati** [comment]
Hi, maybe I've found a way to implement this, but only tested on vaadin 8.x.
Before digging into vaadin 7.x port I want to hear from you if the following code would work for your use case:

1. Open the UI with a **client side** shortcut handler; in this case you don't have to add a `ShortcutListener` to the `ActionManager`
```java
Button btn = new Button("Open another UI (client side)");
EnhancedBrowserWindowOpener opener = new EnhancedBrowserWindowOpener(AboutUI.class)
  .clientSide(true)
  .withShortcut(ShortcutAction.KeyCode.NUM3, ShortcutAction.ModifierKey.CTRL, ShortcutAction.ModifierKey.ALT)
  .doExtend(btn);
```

2. Open with popup blocker workaround. Here there's a bit more code because you should register the `ShortcutListener` in addition to the  `EnhancedWindowOpener` configuration

```java
Button button10 = new Button("Open another UI (popup workaround)");
ShortcutListener shortcutListener = new ShortcutListener("", 
    ShortcutAction.KeyCode.NUM1, new int[]{
    ShortcutAction.ModifierKey.CTRL, ShortcutAction.ModifierKey.ALT
}) {
    @Override
    public void handleAction(Object sender, Object target) {
        button10.click()
    }
};
getActionManager().addAction(shortcutListener);

EnhancedBrowserWindowOpener opener10 = new EnhancedBrowserWindowOpener(AboutUI.class)
    .popupBlockerWorkaround(true)
    .withShortcut(shortcutListener)
    .doExtend(button10);
button10.addClickListener(ev -> opener10.open());
```

**Comment by mcollovati** [comment]
Hi @karthikeyanStar,
I've just pushed the branch **vaadin7-ISSUE-9/shortcuts** with a port of the above functionality.
Be aware that vaadin7 support was targeted to vaadin 7.6.8 so I'm not sure it will work with 7.5.8.

Can you give it a try and let me know if it works?

**Comment by karthikeyanStar** [comment]
Hi Marco,

I have tried, it is not working. The methods are not coming in 7.5.8.

Thanks

**Comment by mcollovati** [comment]
Hi @karthikeyanStar,
forgive the silly question, but have you cloned this repo, locally built the addon on the **vaadin7-ISSUE-9/shortcuts** branch and changed addon version to 0.1.2 in your project?

**Comment by karthikeyanStar** [comment]
Hi Marco,

Added the below dependency in my project and tried.

<dependency>
   <groupId>org.vaadin.addon</groupId>
   <artifactId>enhanced-window-opener</artifactId>
   <version>0.1.1</version>
</dependency>

Thanks

Karthikeyan

**Comment by mcollovati** [comment]
Hi @karthikeyanStar,
**0.1.1** does not have the support for shortcuts and **0.1.2** is not yet published on directory.
I was wondering if you can build the addon on your own to test it before publishing on directory.


As an alternative can you temporarily replace the dependency in your pom with the following one

```xml
<dependency>
  <groupId>com.github.mcollovati.enhanced-window-opener</groupId>
  <artifactId>enhanced-window-opener</artifactId>                                                                                                                                                                         
  <version>vaadin7-ISSUE-9~shortcuts-SNAPSHOT</version>
</dependency>
```
and add the following repository (always on your pom)

```xml
<repositories>
  <repository>
    <id>jitpack.io</id>
    <url>https://jitpack.io</url>
  </repository>
</repositories>
```

This is just to test the addon; if you can confirm it works with this settings I'll publish the version on directory.
Thank you for your patience and for you help on improving this addon

**Comment by karthikeyanStar** [comment]
Hi Marco,

I have tried, but it is not working with the above changes. Can you explain what difference is blocking between Vaadin 7.5.8 vs 7.6.8 version architecture ? Why it is not working in 7.5.8 after 
withShortcut, popupblockerworkaround, clientside options are enabled ??? 

Thanks,

karthikeyan

**Comment by mcollovati** [comment]
Hi,
hope you are well.
Can you please provide a stacktrace or explain more in detail what errors are you facing?

**Comment by mcollovati** [comment]
I just tried to run the demo with vaadin 7.5.8 and it works for me

**Comment by karthikeyanStar** [comment]
Marco,

I am not getting any error with the code, but the opener.open() or button.click() event are called when i debug in eclipse. But the window is not opening. I am using chrome.

Thanks

**Comment by mcollovati** [comment]
Can you please share a sample project so that I can try to replicate the problem?

**Comment by karthikeyanStar** [comment]
Here is the code Marco,

```
Button btnGo = new Button();
btnGo.setCaption("Go");
btnGo.setImmediate(true);


final EnhancedBrowserWindowOpener opener = new EnhancedBrowserWindowOpener(SampleUI.class);
		opener.popupBlockerWorkaround(true);
		opener.clientSide(true);
		opener.withShortcut(callSListener());
		opener.setFeatures("height=700,width=1300,resizable");
		opener.doExtend(btnGo);

btnGo.addClickListener(new Button.ClickListener() {
			@Override
			public void buttonClick(ClickEvent event) {
				opener.open();				
			}
		});
```
		

```
public ShortcutListener callSListener(){
	ShortcutListener shortcutListener = new ShortcutListener("", KeyCode.NUM2, new int[]{ModifierKey.CTRL, ModifierKey.ALT}) {
		    @Override
		    public void handleAction(Object sender, Object target) {
		        btnGo.click();
		    }
		};
               getActionManager().addAction(shortcutListener);
		return shortcutListener;
	}
```
Please share your comments on this....

**Comment by mcollovati** [comment]
`popupBlockerWorkaround` and `clientSide` should not be used together.

If your shortcut is only needed to open the window (no further server side actions) the following solution should work (or almost it `works on my machine` :tm:  also with vaadin 7.5.8   :smiley: )

```
Button btnGo = new Button();
btnGo.setCaption("Go");
btnGo.setImmediate(true);

final EnhancedBrowserWindowOpener opener = new EnhancedBrowserWindowOpener(SampleUI.class);
		opener.clientSide(true);
                opener.withShortcut(KeyCode.NUM2, ModifierKey.CTRL, ModifierKey.ALT)	;
		opener.setFeatures("height=700,width=1300,resizable");
		opener.doExtend(btnGo);
```
Hope this helps you

**Comment by karthikeyanStar** [comment]
Hi Marco,

I am stuck with other work. I will update you soon with this.

**Comment by karthikeyanStar** [comment]
Hi Marco,

Thanks for your patience.

Code is working fine. Please go head and publish the code to the directory.
Share the updated dependency details. 
Further, at present there is no setter method to add the UI class in EnhancedBrowserWindowOpener. If possible, please add that one too, it will be helpful.

Thanks again and sorry for the delayed response.

Karthikeyan

**Comment by mcollovati** [comment]
I'll publish the addon as soon as possibile.

For the UI stuff, can you please open another issue? Also providing sample code on how you want to use it would be helpful

**Comment by mcollovati** [comment]
Addon is published on directory; the maven settings are the following

```xml
<dependency>
   <groupId>org.vaadin.addon</groupId>
   <artifactId>enhanced-window-opener</artifactId>
   <version>0.1.2</version>
</dependency>

<repository>
   <id>vaadin-addons</id>
   <url>http://maven.vaadin.com/vaadin-addons</url>
</repository>
```

Let me know if all works so I can close this issue

**Comment by mcollovati** [comment]
Assuming that latest release work.
I'm going to close this issue.
If there any problems or other feature suggestions feel free to open other issues.

Thanks
